### ISSUE TITLE
Why does sed split files into an array, call replace on each line and rejoin?

### ISSUE THREAD

**Issue by toddw** [shelljs/shelljs#645]
### Node version (or tell us if you're using electron or some other framework):
v6.4.0

### ShellJS version (the most recent version/Github branch you see the bug on):
0.7.6

### Operating system:
Mac OS 10.12.2

### Description of the bug:
Using `shelljs.sed` over a folder full of files, if one of the files is a binary file like a png… sed will corrupt the images.

*More details:*
So say a folder has files of the following types (_.js_, _.css_, _.json_, _.png_, _.jpg_). I wanted to use `sed` to replace a string that occurs only in the js, css, json files. I am 100% certain this string would not exist in the binary formats for the png/jpg files so there should be no harm in just running sed over the entire folder. However, because sed attempts to break every file (including the binary files) into an array, split them up on line breaks and merge it back together the binary files come out the other end completely corrupted.

Looking through the history I see that this file splitting/joining happened in a PR that says it is meant to add support for multiple files. I'm not sure that the splitting of individual files should have been part of that or should be part of the code at all. Is this how normal `sed` works? If so then maybe it is appropriate, if not then I can quickly create a PR to resolve this.

### Example ShellJS command to reproduce the error:

```javascript
var buildFiles = shelljs.ls("build/**").filter(f => f.substr(0, 6) === "build/");
shelljs.sed("-i", /apples/g, "oranges", buildFiles);
```

**Comment by nfischer** [comment]
Splitting files into lines is necessary because `sed` operates line-by-line. Take this example:

```sh
$ cat file.txt
foo foo
foo
$ sed 's/foo/bar/' file.txt
bar foo
bar
$ sed 's/foo/bar/g' file.txt
bar bar
bar
```

A `sed` replacement without a `/g` needs to replace the first instance on every line, and only the first instance.

The PR title may be a bit cryptic. The PR was intended to make `sed` conform to unix behavior, the main feature being multiple file arguments.

If `sed` really is corrupting binary files, that probably is a bug (I can't imagine that unix `sed` would do that). @toddw does the following code also corrupt the binary file?

```js
shell.cat('path/to/pic.png').to('output.png');
shell.exec('diff path/to/pic.png output.png'); // check the files are the same
```

**Comment by toddw** [comment]
@nfischer Running your suggested test, output.png was corrupted and the files were not the same.

**Comment by nfischer** [comment]
Huh, very strange. It looks like there's something wrong with the call to fs.readFileSync()

**Comment by freitagbr** [comment]
This is because we're opening files with the `'utf8'` encoding:
```js
cat += fs.readFileSync(file, 'utf8');
```
.png files are not utf8, so the data gets all garbled when node tries to read it as utf8. This is a sane default in most cases, though we should probably allow the user to pick the encoding if they need to.
This is because we're opening files with the `'utf8'` encoding:
```js
cat += fs.readFileSync(file, 'utf8');
```
.png files are not utf8, so the data gets all garbled when node tries to read it as utf8. This is a sane default in most cases, though we should probably allow the user to pick the encoding if they need to.

**Comment by freitagbr** [comment]
Alternatively, we could add a "raw" option that switches to using `Buffer`s. `fs.readFileSync` and `fs.writeFileSync` would work as expected, and the user wanted to see the buffer as a string, they could use the `.toString` method.

**Comment by nfischer** [comment]
@freitagbr any way we can detect this behind the scenes? It'd be great if we didn't need to worry users with this.

It seems pretty lame that `shell.cat('input.png').to('output.png')` doesn't produce a copy.

**Comment by toddw** [comment]
Would something like this be useful here? https://github.com/sindresorhus/file-type

**Comment by freitagbr** [comment]
Actually, I think the best thing to do would be to call `readFileSync` without an encoding; let it read the raw data. This behavior seems more in-line with `cat`.

For example, take any file (text or binary), and `cat <file>` should produce the same output as `process.stdout.write(fs.readFileSync('<file>'))`.

**Comment by toddw** [comment]
Yeah, that would be good for cat, I guess I was thinking more about sed

**Comment by nfischer** [comment]
Yeah, but internal functions operate on strings. How can we get proper support for non-utf-8 files while still keeping compatibility?

We could keep things as buffers internally, and then use that module to check if utf8 will be ok. But even doing that, we'll have to convert to string-like output for the return value. We could perhaps store buffers inside ShellStrings as well, and use the buffer value in `.to()` and `.toEnd()`. This increases complexity of the ShellStrings though.

---

As for `sed` and other similar functions, we could use `file-type` to check if it's utf8 and then skip over non-text files, stringifying all other files so we can operate on them. This could be an improvement for `grep`, `head`, `tail`, `sort`, `uniq`, and probably others.

**Comment by freitagbr** [comment]
In my testing, converting binary Buffer to a string and back is perfectly fine. The string just contains a lot of characters like "�".

**Comment by nfischer** [comment]
Proposed approach for a solution:

 - [ ] Abstract `fs.readFileSync(fname, encoding)` calls into a new `common.readFileSync(fname)` method, which still returns a string, but which is not garbled by the `'utf-8'` stuff. This probably involves @freitagbr's suggestion of converting between binary buffers and strings.
 - [ ] Replace existing calls to `fs.readFileSync` with the new `common.readFileSync`
 - [ ] Maybe replace `fs.writeFileSync` calls similarly
 - [ ] Add a very small `.png` file to our test resources
 - [ ] Add unit tests that make sure we can successfully read-in and write-out the `.png` file. Verify that `fs.writeFileSync('output.png', fs.readFileSync('test.png', 'utf-8'))` works incorrectly, while `cat('test.png').to('output.png')` works correctly. Also, we should add a `sed('-i', ...)` test case specifically, since that's how this was detected.
 - [ ] Run benchmarks to see what perf regression this creates. If it's unacceptably slow, we can hide this behavior behind an opt-in `shell.config` flag.
 - [ ] Expose `common.readFileSync` to the plugin API

Assigning this over to @freitagbr since he's done some investigation of this. Hopefully this doesn't involve modifying ShellString.

**Comment by Snorlock** [comment]
How is this going?

**Comment by freitagbr** [comment]
I got stuck on this one for a while because I was confused about what the issue really was. It turns out that it's not an encoding issue. We are reading and writing files in `'utf-8'`, which is the correct encoding in this case. As long a you read and write in the same encoding, the files will be fine. If you want to operate on text using a regex, the text needs to be encoded in `utf-8`, so that is the correct choice here. This should be abstracted, so a `common.readFileSync` API is a good idea.

Also, GNU `sed` does not make a distinction between binary and text files; it will happily replace substrings in binary files. It's up to the user to run `sed` on the correct files. Note that BSD `sed` does not operate on files that have non-printing characters or characters outside the user's locale, and many binary files likely have these types of characters.

The real issue here is that we are splitting files using this regex: `/\r*\n/`. Basically, split the file on any number of carriage returns (`\r`) and a newline (`\n`). However, we are only joining the lines back up with a newline. So, if your binary file contains a string of carriage returns and a newline, it will be corrupted.

I think the solution here is to separate lines with the same regex, `/\r*\n/`, but to join with whatever string was found. This means that only the specified strings are replaced, and nothing else.

**Comment by nfischer** [comment]
Hmmm interesting. Yeah, this sounds like a bug. A better regex would be `\r?\n`, but even that may pose problems. For example, what if a binary has a `\r\n` and a `\n` located in different spots?

What does node do on Windows? Does it treat `\n` as the line ending character on Windows? Or does this not apply when writing to a file?

**Comment by freitagbr** [comment]
I think the functional part of `sed` needs a rewrite. Currently, `shell.sed` performs these actions:

1. reads an entire file in memory with `fs.readFileSync`
2. converts buffer to string
3. splits string with `/\r*\n/`
4. replaces the text in each line the given regex
5. rejoins the lines with `'\n'`
6. writes file with `fs.writeFileSync`

I think a better approach is needed. Instead of reading the whole file into memory, we could read smaller chunks into memory, one at a time (like `copyFileSync` in `cp.js`). With the current chunk, chop of from the last newline until the end and save the chunk for later. With the first part of the chunk, split it by `/\n/` (this is more in line with GNU `sed`'s behavior, which doesn't care about carriage returns in my testing). On each line, perform the replace, then rejoin with `'\n'`. Write this to the new file, then read the next chunk. If there was a bit let over from the previous iteration, append the new chunk to it, and repeat. This is more complex than the current implementation but it reduces memory usage with larger files.

Also, I'm pretty sure node just writes whatever you tell it to, regardless of the platform. And I don't think splitting and joining with `'\n'` makes a difference on windows either - as long as the split and join characters are the same.

**Comment by nfischer** [comment]
@freitagbr Doesn't sound like we need an extensive rewrite.

How about we just change it to `.split('\n')` and `.join('\n')`. Wouldn't this fix the issue? This is more consistent with `cat()`. Also, let's do the same fix for `head()`, `tail()`, and `sort()`.

**Comment by freitagbr** [comment]
For now, sure, we don't need an extensive rewrite. But, I still think it would be a good idea to read and write files in chunks. I will file another issue for that.

**Comment by freitagbr** [comment]
Closing, as this should be fixed by #690.
